validation: In case of a continued reindex, only activate chain in the end

mzumsande opened the pull request:
If a user interrupts a reindex while it is iterating over the block files, it will continue to reindex with the next node start (if the `-reindex` arg is dropped, otherwise it will start reindexing from scratch).
However, due to an early call to `ActivateBestChainState()` that only exists to connect the genesis block during 
the original `-reindex`, it wil start connecting blocks immediately before having iterated through all block files.
Because later headers above the minchainwork threshold won't be loaded in this case, `-assumevalid` will not 
be applied and the process is much slower due to script validation being done.

Fix this by only calling `ActivateBestChainState()` here if Genesis is not connected yet (equivalent to `ActiveHeight() == -1`).
Also simplify this spot by only doing this for the active chainstate instead of looping over all chainstates (first commit).

This issue was discussed in the thread below https://github.com/bitcoin/bitcoin/pull/31346#discussion_r1856824817, the impact on assumevalid was found by l0rinc.

The fix can be tested by manually aborting a `-reindex` e.g. on signet and observing in the debug log the order in which blockfiles are indexed / blocks are connected with this branch vs master.

mzumsande committed:
validation: Don't loop over all chainstates in LoadExternalBlock

This simplifies the code. The only reason to call ActivateBestChain()
here is to allow the main init thread to finish startup in a case of
-reindex. In this situation no second chainstate can exist anyway
because -reindex would have deleted any snapshot chainstate earlier.

This could change behavior slightly if -loadblocks was used when there is a
snapshot chainstate. In this case, there is no reason to call
ActivateBestChain() for that chainstate here - it will be called in
ImportBlocks() after all blocks have been indexed.

mzumsande committed:
validation: fix issue with an interrupted -reindex

If a reindex was interrupted while it was iterating
through the block files, genesis will already be connected
when the reindex resumes at the next startup.
In this case, a call to ActivateBestChainState() is not only unnecessary,
but it would connect multiple blocks without applying
-assumevalid, which is much slower.
This is because assumevalid requires us to have a header above
the minimum chainwork, but that header is unknown to us if it's in
a later blockfile not indexed yet.

l0rinc commented:
@mzumsande, do you happen to know why setting `assumevalid` to [block 859999](https://mempool.space/block/0000000000000000000196a7111c7aa3368af5a803a81c7bfb32860100dfd9c7) works as expected (it is found in the `860001` blocks in [`m_blockman.m_block_index`](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2496)), so it's correctly disabling script verification, but setting it to [block 860002](https://mempool.space/block/000000000000000000020c0ad748d78767d986e881259ba6de0a6e62cf946cf1) (we're currently at `873543`) enables `fScriptChecks` from the very beginning since for some reason (i.e. setting a later block results in having earlier validation).

l0rinc commented:
I have tested 570af72, did an IBD to a clean `datadir`, stopped the process at ~20k blocks, restarted it with a `-reindex`: `m_blockman.m_block_index` contains the previous `25921` blocks only, so `m_chainman.AssumedValidBlock()` is not found among them, so script validation gets turned on at `height=1` (same behavior as on `master`).

mzumsande commented:
> @mzumsande, do you happen to know why setting `assumevalid` to [block 859999](https://mempool.space/block/0000000000000000000196a7111c7aa3368af5a803a81c7bfb32860100dfd9c7) works as expected (it is found in the `860001` blocks in [`m_blockman.m_block_index`](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2496)), so it's correctly disabling script verification, but setting it to [block 860002](https://mempool.space/block/000000000000000000020c0ad748d78767d986e881259ba6de0a6e62cf946cf1) (we're currently at `873543`) enables `fScriptChecks` from the very beginning for some reason (i.e. setting a later block results in having earlier validation).

I can't reproduce this. Assuming it's not related to reindex, on an empty datadir after header sync, whether I start with 
`-assumevalid="0000000000000000000196a7111c7aa3368af5a803a81c7bfb32860100dfd9c7"`
or
`-assumevalid="000000000000000000020c0ad748d78767d986e881259ba6de0a6e62cf946cf1"`
`fScriptChecks` is set to 0.

> I have tested https://github.com/bitcoin/bitcoin/commit/570af7281e21d51852b706e891835758d3d7161f, did an IBD to a clean datadir, stopped the process at ~20k blocks, restarted it with a -reindex: m_blockman.m_block_index contains the previous 25921 blocks only, so m_chainman.AssumedValidBlock() is not found among them, so script validation gets turned on at height=1 (same behavior as on master).

Did you do this on mainnet? That would be because `-reindex` deletes the block tree db, but you don't have the full blocks for most of the headers yet. So after reindexing all of the block files you have, you don't have the headers for the assumevalid block, which is not applied as a result.
This is a general issue if you do reindex with a datadir that doesn't have blocks above the minchainwork threshold in it - this PR doesn't solve this.

sedited reviewed (commented):
Concept ACK - I was somehow under the impression that this was done on purpose to guarantee the validity of the already done work before continuing, but could never come up with a really good reason of why this should be the case - and I guess there really isn't one!

Reading through this code again recently makes me wish we'd have better handling for the genesis block in general. It feels weird that it is only activated in the first place during a fresh startup once `ImportBlocks` is called. It is also inconvenient for kernel users, since they need to manually call `ActivateBestChain` on startup (like done in `bitcoin-chainstate`). I wonder if it would be better to move loading it into load chainstate, for example like I did [here](https://github.com/TheCharlatan/bitcoin/commit/c8540a33ba8dc60c5b4f88f7c6ff3ccd1739598a) (though that patch is pretty hacky with hardcoding the flat file position and making the registered validation interfaces miss the first notification and has some other implications).

sedited reviewed (approved):
ACK 570af7281e21d51852b706e891835758d3d7161f

ryanofsky commented:
Concept ACK. These changes seem clearly better than current code, though the fix here seems like a workaround for a workaround for a workaround. IIUC, the original problem leading to this complexity was a crash in some code that assumed the genesis block existed, so a wait was added in #5243 to prevent that crash (first workaround instead of just making code handle the state and not crash). Then this wait apparently prevented gui from showing reindex progress so an ActivateBestChain call was added in #7917 to connect the genesis block and bypass the wait (second workaround, which can connect too many blocks after reindexing is interrupted or when -loadblock is used, leading to this bug). Now a third workaround is being implemented to prevent connecting too many blocks in the typical case when reindexing is interrupted and blocks are written in expected order with genesis first. But this is still a workaround, not a solid fix because other blocks besides the genesis block can still be connected in other cases.

It seems to me like things would be simpler and more reliable if we could delete the wait and delete the ActivateBestChain call, but as long as they need to exist, this fix seems strictly better than the current code.

Am wondering if there is some way to write a functional test for this PR. It'd be hard to write a test that interrupts reindexing at the right time, but maybe there could be a test that uses two -loadblock arguments to load the same blocks multiple times and makes sure all the blocks except the genesis blocks are only connected together at the very end.

mzumsande commented:
Thanks for digging up the history!

> Am wondering if there is some way to write a functional test for this PR. It'd be hard to write a test that interrupts reindexing at the right time, but maybe there could be a test that uses two -loadblock arguments to load the same blocks multiple times and makes sure all the blocks except the genesis blocks are only connected together at the very end.

I don't completely understand this - do you mean -loadblock with specially crafted blockfiles in which block appear in a certain order?

> But this is still a workaround, not a solid fix because other blocks besides the genesis block can still be connected in other cases.

Do you know any concrete examples for the latter? I was thinking that this was more of a solid fix because
* `ActivateBestChain()` cannot connect blocks unless they are in the block tree db. 
* we cannot add blocks to the block tree db unless they connect to genesis (`LoadExternalBlockFile` detects out-of-order blocks and saves them for later).

So, if we call ABC directly after calling `AcceptBlock()` for Genesis, it should be impossible for it to connect any other blocks than genesis because no other headers can be in the block tree db - even in some theoretical scenario where the order within the block file is messed up such that Genesis is not at its beginning.

That being said, I agree it would be nice if we could remove this complexity, but I think it can only be done together with removing the wait-for-genesis in `init` - which means making sure that nothing bad happens when we don't have a genesis block after startup.

mzumsande commented:
So it looks me like there are several ideas for further improvements in this general area:
 1) Consider deactivating assumevalid during reindex altogether - in case of a crash during IBD when we haven't downloaded received blocks above the assumevalid threshold, we currently don't apply assumevalid since we don't have headers above minchainwork but we also won't request headers from peers until the (slow) reindex has finished.
2) Call ActivateBestChain during first startup with an empty datadir earlier, not only in `ImportBlocks` (main challenge: `-reindex` must still work).
3) Remove the genesis-specific `ActivateBestChain()` call - possible together with the wait-for-genesis condition in init.

ryanofsky commented on src/validation.cpp:
```diff
@@ -5158,15 +5158,8 @@ void ChainstateManager::LoadExternalBlockFile(
                 // Activate the genesis block so normal node progress can continue
                 if (hash == params.GetConsensus().hashGenesisBlock) {
-                    bool genesis_activation_failure = false;
-                    for (auto c : GetAll()) {
```
In commit "validation: Don't loop over all chainstates in LoadExternalBlock" (dff5f81433fa8ce1d292a2f6051204a0d7824d6e)

It would be helpful if commit description could clarify whether this is a change in behavior. The "no second chainstate can exist anyway because -reindex would have deleted any snapshot chainstate earlier" part suggests this doesn't change current behavior. But I think in the case when `-loadblock` is being used, not `-reindex`, a snapshot chainstate might actually be loaded and not calling ActivateBestChain would be a (good?) change.

Would just be good to clarify what this change does in terms of externally observable behavior.

mzumsande replied:
I added a clarification. The only reason to have this call here is to not block init, and in the `-loadblock` / `assumeutxo` case (which, by the way, I'm not sure anyone has ever tested) this is not an issue / ActivateBestChain() will be called for both chainstates at the end of `ImportBlocks()`.

ryanofsky commented on src/validation.cpp:
```diff
@@ -5157,7 +5157,10 @@ void ChainstateManager::LoadExternalBlockFile(
+                // Do this only if genesis isn't activated yet, to avoid connecting many blocks
+                // without assumevalid in the case of a continuation of a reindex that
+                // was interrupted by the user.
+                if (hash == params.GetConsensus().hashGenesisBlock && WITH_LOCK(::cs_main, return ActiveHeight()) == -1) {
```
In commit "validation: fix issue with an interrupted -reindex" (570af7281e21d51852b706e891835758d3d7161f)

It would be helpful if comment below mentioned that ActivateBestChain call will only connect the genesis block, not other blocks because ActivateBestChain on connects blocks which are in the block tree db, and block tree db only contains blocks whose parents are in it.

Otherwise it's not clear why ActivateBestChain is being called to connect just one block and why it wouldn't connect other blocks, or at least it wasn't clear to me.

Alternately if there's are some cases where this check is imperfect and ActivateBestChain could connect extra blocks, it would be good if comment clarified to say that intentions of this code is only to connect genesis block but there are case it could connect other blocks.

mzumsande replied:
I extended the comment a bit.
I think one more theoretical case in which `ActivateBestChain` could have connected extra blocks (but wouldn't anymore after this PR) is  if, for some reason, block files loaded here (whether with `-reindex` or `-loadblock`) would contain the genesis block multiple times.

ryanofsky reviewed (approved):
Code review ACK 570af7281e21d51852b706e891835758d3d7161f. This seems like a clear improvement, but I think in longer term things would be a lot simpler we could delete the wait-for-genesis init loop and drop this special case ActivateBestChain call. 

> > Am wondering if there is some way to write a functional test for this PR. It'd be hard to write a test that interrupts reindexing at the right time, but maybe there could be a test that uses two -loadblock arguments to load the same blocks multiple times and makes sure all the blocks except the genesis blocks are only connected together at the very end.
> 
> I don't completely understand this - do you mean -loadblock with specially crafted blockfiles in which block appear in a certain order?

I was thinking a test could use `-loadblock` with a normal block file with blocks in order, but just specify the same `-loadblock` file twice, so when genesis block is encountered the second time it triggers the ActivateBestChain call in the old code and causes non-genesis blocks to be connected to the chain in the middle of loading the block files. I'm not sure how important having a test for this is though, so maybe this would not be worth it, or wouldn't work for some other reason.

> > But this is still a workaround, not a solid fix because other blocks besides the genesis block can still be connected in other cases.
> 
> Do you know any concrete examples for the latter?

I was just thinking of cases with out of order blocks, but I didn't realize that out of order blocks would not be connected because out of order blocks are not in the block tree db. 

Maybe a case where -reindex-chainstate and -loadblock were used together could still trigger this though, if -reindex-chainstate causes ActiveHeight() to return -1 but doesn't clear the block tree db, and one of the -loadblocks files contains a genesis block, it could cause all the blocks in the block tree db to be connected when the genesis block is encountered.

mzumsande commented:
[570af72](https://github.com/bitcoin/bitcoin/commit/570af7281e21d51852b706e891835758d3d7161f) to [c9136ca](https://github.com/bitcoin/bitcoin/commit/c9136ca90605bbe29f005f538b92ff96ca360a13): Addressed feedback by @ryanofsky

ryanofsky reviewed (approved):
Code review ACK c9136ca90605bbe29f005f538b92ff96ca360a13. Only comments changed since last review. Appreciate the new comments, I think they make a little clearer what things code is trying to do and what things are just side-effects.

sedited reviewed (approved):
Re-ACK c9136ca90605bbe29f005f538b92ff96ca360a13

achow101 commented:
ACK c9136ca90605bbe29f005f538b92ff96ca360a13
